From 14654b4d3098bb1cdf906215279a34e6178cbbb9 Mon Sep 17 00:00:00 2001
From: Artem Lapkin <art@khadas.com>
Date: Thu, 12 Aug 2021 12:21:20 +0800
Subject: [PATCH 50/58] arm64: dts: meson: khadas-vim3: remake simple sound for
 i2s hdmi spdif output

Remake sound device to i2s hdmi spdif output
1) simplify ( remove loop + A & C nodes )
2) enable spdif output to GPIO headers
3) enable i2s

NOTE: if u no need i2s output to GPIO headers comment all &tdmif_b
pinctrl* and assigned-clock*

Signed-off-by: Artem Lapkin <art@khadas.com>
---
 .../boot/dts/amlogic/meson-khadas-vim3.dtsi   | 85 ++++++++++++-------
 1 file changed, 53 insertions(+), 32 deletions(-)

diff --git a/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi b/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
index 98613029c6e4..93878c63726a 100644
--- a/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
+++ b/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
@@ -190,20 +190,20 @@ hdmi_connector_in: endpoint {
 		};
 	};
 
+	spdif_dit: audio-codec-1 {
+		#sound-dai-cells = <0>;
+		compatible = "linux,spdif-dit";
+		status = "okay";
+		sound-name-prefix = "DIT";
+	};
 
 	sound {
 		compatible = "amlogic,axg-sound-card";
 		model = "KHADAS-VIM3";
-		audio-aux-devs = <&tdmin_a>, <&tdmout_a>;
-		audio-routing = "TDMOUT_A IN 0", "FRDDR_A OUT 0",
-				"TDMOUT_A IN 1", "FRDDR_B OUT 0",
-				"TDMOUT_A IN 2", "FRDDR_C OUT 0",
-				"TDM_A Playback", "TDMOUT_A OUT",
-				"TDMIN_A IN 0", "TDM_A Capture",
-				"TDMIN_A IN 3", "TDM_A Loopback",
-				"TODDR_A IN 0", "TDMIN_A OUT",
-				"TODDR_B IN 0", "TDMIN_A OUT",
-				"TODDR_C IN 0", "TDMIN_A OUT";
+		audio-aux-devs = <&tdmout_b>;
+		audio-routing = "TDMOUT_B IN 0", "FRDDR_A OUT 1",
+				"TDM_B Playback", "TDMOUT_B OUT",
+				"SPDIFOUT IN 0", "FRDDR_A OUT 3";
 
 		assigned-clocks = <&clkc CLKID_MPLL2>,
 				  <&clkc CLKID_MPLL0>,
@@ -218,29 +218,9 @@ dai-link-0 {
 			sound-dai = <&frddr_a>;
 		};
 
-		dai-link-1 {
-			sound-dai = <&frddr_b>;
-		};
-
-		dai-link-2 {
-			sound-dai = <&frddr_c>;
-		};
-
-		dai-link-3 {
-			sound-dai = <&toddr_a>;
-		};
-
-		dai-link-4 {
-			sound-dai = <&toddr_b>;
-		};
-
-		dai-link-5 {
-			sound-dai = <&toddr_c>;
-		};
-
 		/* 8ch hdmi interface */
 		dai-link-6 {
-			sound-dai = <&tdmif_a>;
+			sound-dai = <&tdmif_b>;
 			dai-format = "i2s";
 			dai-tdm-slot-tx-mask-0 = <1 1>;
 			dai-tdm-slot-tx-mask-1 = <1 1>;
@@ -249,7 +229,7 @@ dai-link-6 {
 			mclk-fs = <256>;
 
 			codec {
-				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_A>;
+				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_B>;
 			};
 		};
 
@@ -261,6 +241,16 @@ codec {
 				sound-dai = <&hdmi_tx>;
 			};
 		};
+
+		/* spdif or toslink interface */
+		dai-link-4 {
+			sound-dai = <&spdifout>;
+
+			codec-0 {
+				sound-dai = <&spdif_dit>;
+			};
+		};
+
 	};
 
 	wifi32k: wifi32k {
@@ -279,6 +269,35 @@ &clkc_audio {
 	status = "okay";
 };
 
+&tdmif_b {
+	// uncomment next block  enable I2S output to GPIO HEADERS
+	//
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_SCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD1>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_B_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_LRCLK>;
+	assigned-clock-rates = <0>, <0>;
+	/*
+	pinctrl-0 = <&tdm_b_fs_pins>, <&tdm_b_dout0_pins>,
+		    <&tdm_b_sclk_pins>, <&mclk0_a_pins>;
+	*/
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&tdmout_b {
+	status = "okay";
+};
+
+&spdifout {
+	// uncomment next block  enable SPDIF output to GPIO HEADERS
+	/*
+	pinctrl-0 = <&spdif_ao_out_pins>;
+	*/
+	pinctrl-names = "default";
+	status = "okay";
+};
+
 &cec_AO {
 	pinctrl-0 = <&cec_ao_a_h_pins>;
 	pinctrl-names = "default";
@@ -490,6 +509,7 @@ w25q128: spi-flash@0 {
 	};
 };
 
+/*
 &tdmif_a {
 	status = "okay";
 };
@@ -513,6 +533,7 @@ &toddr_b {
 &toddr_c {
 	status = "okay";
 };
+*/
 
 &tohdmitx {
 	status = "okay";
-- 
2.25.1

